query.go

```diff
@@ -785,43 +790,127 @@ func (s *ChainService) prepareCFiltersQuery(blockHash chainhash.Hash,
 	filterType wire.FilterType, options ...QueryOption) (
 	*cfiltersQuery, error) {
 
-	getHeader := s.RegFilterHeaders.FetchHeader
+	_, height, err := s.BlockHeaders.FetchHeader(&blockHash)
+	if err != nil {
+		return nil, fmt.Errorf("unable to get header for start "+
+			"block=%v: %v", blockHash, err)
+	}
 
-	// In order to verify the authenticity of the filter, we'll fetch the
-	// target block header so we can retrieve the hash of the prior block,
-	// which is required to fetch the filter header for that block.
-	block, height, err := s.BlockHeaders.FetchHeader(&blockHash)
+	bestBlock, err := s.BestBlock()
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("unable to get best block: %v", err)
 	}
-	if block.BlockHash() != blockHash {
-		return nil, fmt.Errorf("Couldn't get header for block %s "+
-			"from database", blockHash)
+	bestHeight := int64(bestBlock.Height)
+
+	qo := defaultQueryOptions()
+	qo.applyQueryOptions(options...)
+
+	// If the query specifies an optimistic batch we will attempt to fetch
+	// the maximum number of filters in anticipation of calls for the
+	// following or preceding filters.
+	var startHeight, stopHeight int64
+
+	switch qo.optimisticBatch {
+
+	// No batching, the start and stop height will be the same.
+	case noBatch:
+		startHeight = int64(height)
+		stopHeight = int64(height)
+
+	// Forward batch, fetch as many of the following filters as possible.
+	case forwardBatch:
+		startHeight = int64(height)
+		stopHeight = startHeight + wire.MaxGetCFiltersReqRange - 1
+
+		// We need a longer timeout, since we are going to receive more
+		// than a single response.
+		options = append(options, Timeout(QueryBatchTimeout))
+
+	// Reverse batch, fetch as many of the preceding filters as possible.
+	case reverseBatch:
+		stopHeight = int64(height)
+		startHeight = stopHeight - wire.MaxGetCFiltersReqRange + 1
```

Comment:
Hmm I'm a bit worried about this. I guess the assumption is that by the time we get to do a reverse batch filter fetch, they've most likely already been fetched from a previous forward batch? Otherwise, let's say the reverse batch happens before anything else. On mobile, we expect users to be opening and closing the app multiple times a day, so this would be a pretty big bandwidth hit if we end up fetching 1000 filters unnecessarily every time. Not sure what the best solution is other than lowering the max range for a reverse batch.

Reply:
That's a good point, can totally happen that in most cases we will need only a few of the filters at tip.

I think however that it is a bit premature to be special casing this before we got some real data to back it up. Actually, maybe we should even leave reverse fetching out for now, until we know the impact of using it?

Reply:
I don't think we want to remove it. For Neutrino, we scan backwards when scanning for a historical confirmation. If we don't address the reverse batch range, then we basically depend on the order of `lnd`'s subsystems to ensure a spend registration happens before a confirmation registration, which doesn't seem like a reasonable option to me.

FWIW, currently it seems like spend registrations usually happen before confirmations due to starting the `ChannelArbitrator`s, but this might not always be the case in the future.

Reply:
Solution for now seems to be for the caller to not request an optimistic batch if the range of filters to scan is known to be small?